org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java

```diff
@@ -616,6 +616,15 @@ private boolean hasSingleAttributeInFullDoc(DOMElement element) {
 			Position endPosition = this.textDocument.positionAt(this.endOffset);
 			Range r = new Range(startPosition, endPosition);
 			List<TextEdit> edits = new ArrayList<>();
+
+			// check if format range reaches the end of the document
+			if (this.endOffset == this.textDocument.getText().length()) {
+				this.xmlBuilder.trimTrailingNewlines();

```

Comment:
I think we should do that only if  if `options.isTrimFinalNewlines()` is equals to true

Reply:
I'm not sure if I understand, are you suggesting:

```
if (options.isTrimFinalNewlines() && this.endOffset == this.textDocument.getText().length()) {
    this.xmlBuilder.trimTrailingNewlines();
```
?

Reply:
No I mean 

```java
if (this.endOffset == this.textDocument.getText().length()) {
    if (options.isTrimFinalNewlines()) {
        this.xmlBuilder.trimTrailingNewlines();
    }
    ...
```

Reply:
@angelozerr ok thank you.

Just to clarify, the `xml.format.preservedNewlines` should only be for newlines in between nodes? Therefore `xml.format.preservedNewlines` shouldn't affect the newlines at the end of an XML document?

Reply:
I discovered that the root cause of https://github.com/redhat-developer/vscode-xml/issues/196 is a small bug with how `xml.format.preservedNewlines` is being used

Reply:
I think it shouldn't affect the newlines at the end of the document because in the description:
![image](https://user-images.githubusercontent.com/20326645/80006529-fc3a6b00-8492-11ea-95fa-49417ddbb185.png)
it mentions "new lines that separate tags"

Reply:
To be honest with you, I don't know this formatting option. This newline have started with the issue https://github.com/eclipse/lemminx/issues/67 So I suggest that you play with HTML language server to see in action the preserve newlines behavior and try to do the same thing for XML

> I discovered that the root cause of redhat-developer/vscode-xml#196 is a small bug with how xml.format.preservedNewlines is being used

Is it easy to fix it?

Reply:
Yes, my fix was to not use `xml.format.preservedNewlines` if the current textnode was the last node of the document:
https://github.com/eclipse/lemminx/blob/4fd0d5c99f8373847a7242340f07e48e0620c2d9/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLBuilder.java#L240